prophy/tests/test_compare.py

```diff
@@ -0,0 +1,204 @@
+                       ("value", prophy.array(_type, bound="len"))]
+    return X
+
+
+def get_array_bound_class2(_type):
+    class X(prophy.with_metaclass(prophy.struct_generator, prophy.struct)):
+        _descriptor = [("len", prophy.i32),
+                       ("value", prophy.array(_type, bound="len"))]
+    return X
+
+
+def get_enum_class1():
+    class X(prophy.with_metaclass(prophy.enum_generator, prophy.enum)):
+        _enumerators = [
+            ("x", 1),
+            ("y", 2),
+        ]
+    return X
+
+
+def get_enum_class2():
+    class X(prophy.with_metaclass(prophy.enum_generator, prophy.enum)):
+        _enumerators = [
+            ("x", 1),
+            ("y", 3),
+        ]
+    return X
+
+
+def get_enum_class3():
+    class X(prophy.with_metaclass(prophy.enum_generator, prophy.enum)):
+        _enumerators = [
+            ("x", 1),
+            ("z", 3),
+        ]
+    return X
+
+
+def get_fixed_bytes():
+    class X(prophy.with_metaclass(prophy.struct_generator, prophy.struct_packed)):
+        _descriptor = [("value", prophy.bytes(size=5))]
+    return X
+
+
+def get_bound_bytes():
+    class X(prophy.with_metaclass(prophy.struct_generator, prophy.struct_packed)):
+        _descriptor = [("value_len", prophy.u32),
+                       ("value", prophy.bytes(bound="value_len"))]
+    return X
+
+
+def get_shift_bound_bytes():
+    class X(prophy.with_metaclass(prophy.struct_generator, prophy.struct_packed)):
+        _descriptor = [("value_len", prophy.u8),
+                       ("value", prophy.bytes(bound="value_len", shift=2))]
+    return X
+
+
+def get_limited_bytes():
+    class X(prophy.with_metaclass(prophy.struct_generator, prophy.struct_packed)):
+        _descriptor = [("value_len", prophy.u32),
+                       ("value", prophy.bytes(size=5, bound="value_len"))]
+    return X
+
+
+def get_greedy_bytes():
+    class X(prophy.with_metaclass(prophy.struct_generator, prophy.struct_packed)):
+        _descriptor = [("value", prophy.bytes())]
+    return X
+
+
+@pytest.mark.parametrize('_type', [
+    prophy.i8,
+    prophy.i16,
+    prophy.i32,
+    prophy.i64,
+    prophy.u8,
+    prophy.u16,
+    prophy.u32,
+    prophy.u64,
+    prophy.r32
+])
+def test_compare_descriptors_scalars(_type):
+    x = get_integer_class1(_type)
+    y = get_integer_class1(_type)
+    assert x == y
+
+
+@pytest.mark.parametrize('_type', [
+    prophy.i8,
+    prophy.i16,
+    prophy.i32,
+    prophy.i64,
+    prophy.u8,
+    prophy.u16,
+    prophy.u32,
+    prophy.u64,
+    prophy.r32
+])
+def test_compare_descriptors_arrays(_type):
+    a, b = get_array_bound_class1(_type), get_array_bound_class1(_type)
+    c, d = get_array_bound_class2(_type), get_array_bound_class2(_type)
+    e, f = get_array_class1(_type), get_array_class1(_type)
+    g, h = get_array_class2(_type), get_array_class2(_type)
+    i, j = get_array_class3(_type), get_array_class3(_type)
+
+    assert a == b
```

Comment:
That's not quite true. `a` is a different type than `b`. They can be _similar_, but they're not `==`.
```
(Pdb++) def Foo(): pass
(Pdb++) def Bar(): pass
(Pdb++) Foo == Bar
False
```

Maybe a custom free function to compare similar types would be in place? `__eq__` operator seems to have a different meaning in Python.

Reply:
I believe meaning of `__eq__` is just `equal` and is free to be implemented or overridden, according to needs and concept.
When it comes to classes, situation is quite special.
Two classes, by concept, should mean two different things, even if their definition at a time is same (their specific attrs are not used in the scope). So default `__eq__` should return False according to my understanding, and it is.
Usually, people deal with one source of class definition, and they never meet situation where different definitions do actually mean same thing and they never have to detect it.
Our case is special - we have multiple sources of class definitions, which classes do mean exactly the same, and we need to know that. 
In my opinion, using `__eq__` is our case is absolutely correct.

Reply:
In my opinion use of `__eq__` is correct and I cannot find any reason to avoid it. If `type` would implement `__eq__` it could be fishy to override it, but it does not. It's not a change of default behavior of `type`, but its supplementation of its missing functionality.

Reply:
It implements `__eq__`:
```
In [9]: type.__eq__
Out[9]: <slot wrapper '__eq__' of 'object' objects>
```
https://stackoverflow.com/questions/35709339/what-is-the-source-code-of-hash-and-eq-of-object-in-python
https://github.com/python/cpython/blob/master/Objects/typeobject.c#L3877
If it hadn't implemented it, you couldn't define two custom types and compare them like I did ^^.

> Usually, people deal with one source of class definition, and they never meet situation where different definitions do actually mean same thing and they never have to detect it.
> Our case is special - we have multiple sources of class definitions, which classes do mean exactly the same, and we need to know that.
> In my opinion, using __eq__ is our case is absolutely correct.

Legal, but immoral :)
`__eq__` is just an operator which can be replaced with any free function. I'm arguing this operator has specific meaning in Python for types and it's not wise to override it. Custom free function is explicit, `==` for two types, on the other hand, suggests to majority of Python developers that it's just an identity check - which in this case is a wrong suggestion.

Anyway, if you still prefer `__eq__` - let's go with it.